src/store/addressesSlice.ts

```diff
@@ -172,84 +111,118 @@ export const fetchAddressesTransactionsNextPage = createAsyncThunk(
   }
 )
 
-export const mainAddressChanged = createAsyncThunk(
-  `${sliceName}/mainAddressChanged`,
-  async (payload: Address, { getState, dispatch }) => {
-    const newMainAddress = payload
+export const addressesFromStoredMetadataInitialized = createAsyncThunk(
+  `${sliceName}/addressesFromStoredMetadataInitialized`,
+  async (_, { getState, dispatch, rejectWithValue }) => {
+    const state = getState() as RootState
 
-    dispatch(loadingStarted())
+    const { metadataId, mnemonic } = state.activeWallet
 
-    const state = getState() as RootState
-    const mainAddress = state.addresses.entities[state.addresses.mainAddress]
+    if (metadataId && mnemonic) {
+      dispatch(loadingStarted())
+
+      const { masterKey } = await walletImportAsyncUnsafe(mnemonicToSeed, mnemonic)
+      const addressesMetadata = await getAddressesMetadataByWalletId(metadataId)
+
+      console.log(`👀 Found ${addressesMetadata.length} addresses metadata in persistent storage`)
 
-    if (mainAddress && mainAddress.hash === newMainAddress.hash) {
-      throw 'Main address is already set to this address'
+      const addresses = addressesMetadata.map(({ index, ...settings }) => ({
+        ...deriveAddressAndKeys(masterKey, index),
+        settings
+      }))
+
+      dispatch(addressesAdded(addresses))
+      await dispatch(addressesDataFetched(addresses.map((address) => address.hash)))
```

Comment:
I've started to reflect a bit on the past tense being used for dispatched actions. We start to use a lot of them, and I realise that it's sometimes a bit counterintuitive & too verbose for me.

Could we take some time soon for you to explain to me (again?) why you choose not to use the present form for this?

An initial thought that may make sense to me:
- use present for synchronous actions (like `flushPin` for instance)
- use past tense for async actions that may be awaited (like `await addressesDataFetched`). 

Not 100% sure either.

Reply:
## Redux recommended style guide

So, when I started learning about Redux and introducing it to the mobile wallet, I came across this recommendation and started using it directly:

> However, **we recommend trying to treat actions more as "describing events that occurred", rather than "setters".**
> https://redux.js.org/style-guide/#model-actions-as-events-not-setters

Their arguments are:
> Treating actions as "events" generally leads to:
> 1. more meaningful action names
> 2. fewer total actions being dispatched
> 3. a more meaningful action log history

As the style guide continues:

> **Actions should be written with meaningful, informative, descriptive type fields.** Ideally, you should be able to read through a list of dispatched action types, and have a good understanding of what happened in the application without even looking at the contents of each action.

## Personal opinion

Funnily enough, I have the complete opposite opinion of you, but I think I have good arguments to support it.

First of all, all of our current reducers/actions as well as our async thunks are written in past forms, as events that happened.

You proposed (tentatively) that reducers/actions should be written in the present tense and async thunks in past.

I propose that reducers/actions are written in past tense and the async thunks in present, and here's my reasoning:

Regarding argument (1), I cannot see how treating actions as events leads to more meaningful action names. The word "action" to me indicates that something is a "command", like "startLoading" or "fetchPrice". Also, when going through the code, it makes more sense to read `dispatch(startLoading())` rather than `dispatch(loadingStarted())` because my mental model around reducers/actions is (probably) wrong. I see them as functions, whereas Redux recommends to see them as events. So regarding this one, I actually don't care. I am fine to go with the recommended approach and tread them as events, using the past tense (especially due to point 3 though, not point 1).

Regarding argument (2), I also don't see any benefit 😅 I've read a couple times that we should avoid dispatching multiple actions one after the other and I try my best to reduce this and improve the code whenever I have a better idea. But there are a few times that I do dispatch 2 or 3 or sometimes 4 actions one after the other. I don't see how changing the tense of the reducer/action name changes anything.

Regarding argument (3) however, this is where I have a stronger opinion. Looking at the debugger, reading the list of "normal" actions that took place and their effect on the state makes a lot of sense (see the following example with `settings/generalSettingsChanged` followed by the actual change):

<img width="816" alt="image" src="https://user-images.githubusercontent.com/1579899/207364956-0bc852aa-df18-4db3-80b8-4f6106edbe55.png">

In the screenshot above I can see that "Oh, I see that the `price/statusChanged` to `updating` and then the `settings/generalSettingsChanged` from `'USD'` to `'CHF'`. It almost reads as natural English language.

What I have a slight problem with, however, is the async thunks (which is the reason I believe why you initially started this thread). See following example:

<img width="666" alt="image" src="https://user-images.githubusercontent.com/1579899/207366254-f1ecf966-faa8-489e-b565-abb1a76e6488.png">

Reading `price/priceUpdated/pending` does not make much sense to me. It would make more sense and it would be more intuitive to read `price/updatePrice/pending`, then `price/statusChanged`, and finally `price/updatePrice/fullfiled`. Same for `addressDataFetched` vs `fetchAdddressData`:

<img width="998" alt="image" src="https://user-images.githubusercontent.com/1579899/207367029-f16c7e69-8839-43b7-84e5-352aac1534b1.png">

The examples on async thunks in the RTK seem to support this idea, although they also use present tense for "normal" reducers/actions as well. https://redux-toolkit.js.org/api/createAsyncThunk

Sorry for the long comment, but I think it's worth having this discussion now! Looking forward to your thoughts!

Reply:
Continuing my research on the topic, I think I really like this opinion. It's worth the read: https://medium.com/magnetis-backstage/why-action-is-a-bad-name-for-a-redux-action-68bec375539e

Specifically, this hit a chord in me:

> Actions are a payload of information that sends data from your application to your store. [...] It's the one and only form of communication between the application and your Store. [...] What should my application be telling my store? [...] Most projects show a mix of the two: communication of intent and communication of an event. [...] the application should inform the Store of what's happening and the Store will know what to do [...] actions are a mean of communication with the outside world. [...] It’s not action, it’s a reaction. It’s not giving orders, it’s describing behavior. It’s not imperative, it’s declarative. And that’s why Action is a bad name for a Redux Action.

I think I have already introduced many anti-patterns in our code-base, probably because I am thinking of actions as functions...

Reply:
My only concern with treating actions as functions and not as events is that we are being led into redux antipatterns... Everywhere in the fundamentals of redux I read:

> The actions, the events that occur in the app based on user input, and trigger updates in the state

> You can think of an action as an event that describes something that happened in the application.

> You can think of dispatching actions as "triggering an event" in the application. Something happened, and we want the store to know about it. Reducers act like event listeners, and when they hear an action they are interested in, they update the state in response.

https://redux.js.org/tutorials/fundamentals/part-2-concepts-data-flow

I would like to continue the research on "why treating actions as functions and not as events leads to antipatterns".

Reply:
What I mentioned just above, btw, is also in line with your comment above: https://github.com/alephium/mobile-wallet/pull/113#discussion_r1046016355

> It's a bit weird to me that `startScan` is dispatching `addressesDiscovered` instead of `addressDiscoveryStarted` 🤔
What's the rational behind this?

I think you're absolutely right. `startScan` should dispatch `scanStarted` (or `addressDiscoveryStarted`, whatever) and let the `addressDiscoverySlice` react to this event by starting the address discovery algo.

_Continues research..._